src/NetTopologySuite/NtsGeometryServices.cs

```diff
@@ -115,6 +116,9 @@ public NtsGeometryServices(CoordinateSequenceFactory coordinateSequenceFactory,
             DefaultSRID = srid;
             GeometryOverlay = geometryOverlay ?? throw new ArgumentNullException(nameof(geometryOverlay));
             CoordinateEqualityComparer = coordinateEqualityComparer ?? throw new ArgumentNullException(nameof(coordinateEqualityComparer));
+
+            _wktReader = new WKTReader(this);
+            _wkbReader = new WKBReader(this);
```

Comment:
All the changes in this file really smell:

- Members are all supposed to be read-only, since this is intended to be a singleton shared by multiple concurrent threads.
- `WKTReader` and `WKBReader` depend on `NtsGeometryServices`, so why does `NtsGeometryServices` also depend on them?
- Why are we just silently ignoring when the new properties get set to `null`, instead of throwing `ArgumentNullException`?

I feel that this file **really** shouldn't need to change in order to enable curved geometries.  If the way to make that happen is to bring them into this project, then let's do that.

Reply:
> Unfortunatly this is necessary because otherwise `Geometry.ToText` and `Geometry.ToBinary` would fail on `GeometryCollection`s containing any curve geometry.

Would it still be necessary if we brought curved geometries into this project?

Reply:
No, but I'm not very comfortable with it.

Reply:
Additionally, having a `WK[T|B]Writer` configureable at the `NtsGeometryServices` instance let us overcome the limitation that we output three dimensions at most. I don't know if constructing a `WK[T|B]Writer` any time `To[Text|Binary]` is called is such a good design decision anyway.

Reply:
Can we find a way to do this that doesn't require making subclasses of these types, at least?  The thing that's actually **needed** here is a way for external code to override what happens when the system sees specific geometry types.

So perhaps we should develop a model where they can inject something like:
```csharp
class MyFancyWkbHandler : WkbHandler
{
    protected override bool TryRead(WKBGeometryTypes geometryType, CoordinateSystem cs, int srid, BiEndianBinaryReader wkbReader, out Geometry geometry)
    {
        // if geometryType doesn't match what we care about, return false
        // else, read the rest of it from wkbReader to geometry and return true
    }

    protected override bool TryWrite(BinaryWriter writer, bool includeSRID, Geometry geometry)
    {
        // if geometryType doesn't match what we care about, return false
        // else, write it to writer and return true
    }
}

class MyFancyWktHandler : WktHandler
{
    // same concept here
}

static void SomeSetupMethod()
{
    NtsGeometryServices.Instance.Register(new MyFancyWkbHandler());
    NtsGeometryServices.Instance.Register(new MyFancyWktHandler());
}
```
And then we would try the registered handlers first before moving onto the default ones?

Reply:
Adding special handler classes is not simple because code to read e.g. coordinate sequences is to be needed for those to.
I think overriding the non-sealed reader/writer classes is the right thing to do.
I hardened NtsGeometryServices by adding a constructor that takes reader/writer creation functions.